### ISSUE TITLE
folder convention for shared directives, services, etc.

### ISSUE THREAD

**Issue by mg1075** [meanjs/mean#607]
Especially with regard to 0.4, I cannot tell what strategy is appropriate when creating a directive or a factory that is not specific to any one module but that might end up being used in multiple modules. 

From the perspective of folder structure, if I made a custom directive potentially applicable to multiple modules, would I create a "directives" folder under modules/core/client/ ?

**Comment by dancancro** [comment]
@mg1075 I don't have the answer for you, but I did a survey of approaches that might interest you.

[Click here](https://docs.google.com/spreadsheets/d/1tGPTpRGQNUsQ117wUI_r5K40FvQt1PFcMZTmfZ7zAPM/edit?usp=sharing)

Rows 76 and 77 answer the question for directives which I'm calling "client-side, app-scope html extensions" and "client-side, app-component-scope html extensions".

Of the approaches I studied, two included examples of client-side, app-scope html extensions:

* ./CC.Web/app/services/directives.js
* ./client/app/directives/productImgSrcDirective.js

and five included examples of client-side, app-component-scope html extensions:

* ./src/client/app/layout/ht-sidebar.directive.js
* ./frontend/app/modules/auth/directives/AccountChooserDirective.js
* ./client/components/mongoose-error/mongoose-error.directive.coffee
* ./client/app/components/navbar/navbar.directive.html
* ./src/common/plusOne/plusOne.js

If anyone reading this would like to fill in a column for mean.js, let me know.

**Comment by mg1075** [comment]
@dancancro - Thanks for sharing. 
With the mean.js 0.4 branch, I don't know if this is acceptable, but adding a folder under "modules" called "_common" seems like a potential approach. 
../modules/_common
../modules/_common/directives
../modules/_common/services
../modules/_common/filters

But this seems not quite right in the sense that "_common" is not a module like the real modules under the modules folder. Putting them under "public" would not be good because there is the potential for server-related code under the _common sub-folders.

**Comment by mleanos** [comment]
When I first started working with MEAN stack, I had the same design consideration. I ended up putting these types of directives/services inside the core/client folder structure. 

However, I've built directives, and services, that mostly pertain to one module but needed to be used in a few places by other modules. I just put these in their respective "module"/client folder. 

The good thing, is that it's not detrimental where you put them. Since any of your angular modules will be able to use them; just as long as they are registered to one of your modules.

I know we all want to keep our projects clean, and well organized. For me using the core module for purposes of these types of client-side components has proven to be easy to manage.
